### ISSUE TITLE
span elements are included in NBG(R_i)

### ISSUE THREAD

**Issue by palemieux** [w3c/imsc#571]
The quantity NBG(R<sub>i</sub>) counts the number of `tts:backgroundColor` attributes specified  `span` elements. 

The application of background on a span element is however more akin to the application of a color or font style.

`span` should be excluded from NBG(R<sub>i</sub>) computation, and `tts:backgroundColor` should instead be included in the list of glyph properties at https://www.w3.org/TR/ttml-imsc1.1/#paint-text.

**Comment by nigelmegitt** [comment]
@palemieux please could you explain more about how this would work? I don't understand the reasoning at the moment: 

The paint text model is based on glyph copying, and the glyphs are copied _after_ the background has been drawn. The background is not the same for every glyph with the listed style properties, even if `tts:backgroundColor` is included, because it does not always have the same size. It may be extended either in the inline or the block progression direction by use of `itts:fillLineGap` or `ebutts:linePadding`.

Even if the background _were_ the same size, the rendering performance factor for drawing a background would need to be taken into account, and presumably would be much less than for drawing the glyph: in this proposal what change would be made to the values of <code>Ren(G<sub>i</sub>)</code>?

**Comment by palemieux** [comment]
@nigelmegitt Do you agree that the current complexity measure, where a full region redraw is assumed on every span-specified `tts:backgroundColor`, does not make sense, right? In other words, complexity for span-specified `tts:backgroundColor` should scale with character area and not region area.

**Comment by nigelmegitt** [comment]
@palemieux it depends what you need to achieve. I do agree that real implementations are unlikely to redraw all of the pixels in a region _n_ times where _n_ is the number of background colours specified for active children of that region. At least, it's very much a worst case scenario. I could imagine some actually do have event loops that might cause this many redraws, if they update some background painting map every time they encounter a new background rectangle.

But as a proxy measure of complexity, which is easy to compute, I think it is okay to assume this worst case scenario, generate _some number_ (which is all the HRM does, after all), and then set thresholds based on the numbers that real world content requires. 

Sure, we could go to a huge amount of further detail, computing the areas of each of the elements selected into the region, and multiplying the number of pixels by the number of times each element gets drawn etc. 

Let me ask you a more meta question back: what is the motivation behind increasing the level of precision? Is there some real world content that colours a lot of span backgrounds differently and falls foul of the HRM?

**Comment by palemieux** [comment]
@nigelmegitt Below is the kind of document I have in mind:

```
<?xml version="1.0" encoding="UTF-8"?>
<tt xml:lang="en"
    xmlns="http://www.w3.org/ns/ttml"
    xmlns:tts="http://www.w3.org/ns/ttml#styling"
    xmlns:ebutts="urn:ebu:tt:style">
  <head>
    <layout>
      <region xml:id="r1"
              tts:origin="10% 10%"
              tts:extent="80% 80%"
              tts:color="white"
              tts:textAlign="center"
              tts:backgroundColor="transparent"
              tts:showBackground="whenActive"
              tts:displayAlign="after"
              ebutts:linePadding="0.5c"/>
    </layout>
  </head>
  <body region="r1">
    <div>
      <p begin="0s" end="0.3s">
        <span tts:backgroundColor="black">Hello my name is</span><br/>
        <span tts:backgroundColor="black">Gaspard Hunt</span><br/>
        <span tts:backgroundColor="black">and I will <span tts:backgroundColor="red">rock</span> your world...</span>
      </p>
      <p begin="0.3s" end="2s">
        <span tts:backgroundColor="black">Gaspard Hunt</span><br/>
        <span tts:backgroundColor="black">and I will <span tts:backgroundColor="red">rock</span> your world...</span><br/>
        <span tts:backgroundColor="black">But first let's take a selfie.</span>
      </p>
    </div>
  </body>
</tt>
```

The `<p begin="0.3s" end="2s">` has four elements with non-transparent `tts:backgroundColor`. Each one triggers a full redraw of the entire region, whose area is 64% of the root container region, even though each element is much smaller than the entire region. As a result the HRM draw time is 0.349s, which means that the previous `<p>` can be on-screen for no more than 0.349s.

The region is transparent and large so that the document can use a single region with no risk of overflow.

**Comment by nigelmegitt** [comment]
Thanks for that example @palemieux . It clarifies things. I'm still thinking about it, and want to share what I'm mulling over:
* My [assertion that the HRM is merely an expression of complexity](https://github.com/w3c/imsc/issues/571#issuecomment-853974170) is not really correct - the way that it defines a double buffer duration means that this is, to me, a more concrete problem.
* The value of BDraw may be too conservative?
* If we were to move to a glyph (or grapheme cluster?) based model for span background colours, then the painting area would not incorporate additional background painting caused by the use of `itts:fillLineGap` and `ebutts:linePadding`; however, adding those in may be considered 
  1. to be a small effect normally and 
  2. to be complex, since a static analysis of the document cannot reveal the number of lines that will be generated at layout time - to do that, something approaching an actual render would be needed.
* NRGA is based on `fontSize * fontSize`, but should probably use `lineHeight * fontSize`  if we are making this change, to take into account the greater area.

**Comment by palemieux** [comment]
@nigelmegitt I agree it is too early to reach a conclusion on this issue, and recommend waiting for more feedback from implementers and users.

**Comment by btsimonh** [comment]
I'm at a loss to understand why every span background would cause a cost of a complete region fill...  I would expect a new buffer per time period, with a single clear and then draw on.
In fact, in real world double buffered implementations, you may expect each buffer to be tagged with it's dirty rectangle, and only that area to need to be cleared - if the buffers are re-used - resulting in a much reduced pixel fill.

_Note on region size: I have generally adopted the use of regions which extend from the top or bottom safe area edge to the required text position.  In practice, as in pierre's example, this means large regions for every title.  Small region sizes implies you know the text extents, and in this world of 'make my text bigger' in renderers, you can't know the final text sizes as the player may change that.  So the best you can do is 'anchor' the text to the position you want it, and encourage the player to expand from there?_

**Comment by nigelmegitt** [comment]
> So the best you can do is 'anchor' the text to the position you want it, and encourage the player to expand from there?

You could alternatively set `tts:overflow="visible"` and set the appropriate value of `tts:displayAlign` if you don't need any other behaviour that depends on the region edges, such as background color.

**Comment by nigelmegitt** [comment]
> I'm at a loss to understand why every span background would cause a cost of a complete region fill... I would expect a new buffer per time period, with a single clear and then draw on.

The HRM is not intended to be a representation of a performance-optimised real world renderer - in fact it probably makes sense to make it in some ways a deliberately sub-optimal model, so that even naive implementations are likely to play back conformant documents acceptably.

**Comment by palemieux** [comment]
> so that even naive implementations are likely to play back conformant documents acceptably.

As it stands, the complexity of drawing the background of a span remains comparable to drawing the entire background of a parent region that fills the root container, even if the font size of the span is 0. I do not think believe this is correct.

**Comment by nigelmegitt** [comment]
> the complexity of drawing the background of a span remains comparable to drawing the entire background of a parent region that fills the root container

Yes, this reflects an _extremely_ naive renderer, for example one that proceeds linearly through rendering, writing directly to a single pixel buffer, and redraws all the background areas every time it encounters something that modifies how those background areas should be drawn.

I'm not saying that the pointer is necessarily at the right place on the dial now; just that moving it all the way to "highly optimised" would also be a mistake.

**Comment by btsimonh** [comment]
painting the whole region for every background colour makes no sense after reading 10.2, which explicitly groups the painting of all background colours before the painting of foregrounds.  
Indeed, in my own renderer, during parse I make a list of all the boxes that need to be painted (whilst noting all the text to be drawn and it's size, 'cos else you don't know the box size), then only when all positioning is known, paint all the boxes, then the outlines, then the foregrounds.   It's actually really important to separate these things, else a background/outline will overlay a foreground, and that is not good....
Indeed, I would argue for adding painting of outlines as an extra step in 10.2, with an extra render cost - because in reality that's how you have to do it.
